modules/ghlint-rules/src/main/kotlin/net/twisterrob/ghlint/rules/MissingKnownActionPermissionsRule.kt

```diff
@@ -0,0 +1,86 @@
+import net.twisterrob.ghlint.model.WorkflowStep
+import net.twisterrob.ghlint.model.asEffectivePermissionsSet
+import net.twisterrob.ghlint.rule.Example
+import net.twisterrob.ghlint.rule.Issue
+import net.twisterrob.ghlint.rule.Reporting
+import net.twisterrob.ghlint.rule.report
+import net.twisterrob.ghlint.rule.visitor.VisitorRule
+import net.twisterrob.ghlint.rule.visitor.WorkflowVisitor
+
+public class MissingKnownActionPermissionsRule : VisitorRule, WorkflowVisitor {
+	override val issues: List<Issue> = listOf(MissingRequiredActionPermissions)
+
+	override fun visitWorkflowUsesStep(reporting: Reporting, step: WorkflowStep.Uses) {
+		super.visitWorkflowUsesStep(reporting, step)
+
+		val expectedPermissions = KnownActionPermissions[step.uses.action] ?: return
+		val definedPermissions = step.parent.permissions?.asEffectivePermissionsSet()
+				?: step.parent.parent.permissions?.asEffectivePermissionsSet()
+				?: return
+
+		val remaining = expectedPermissions.minus(definedPermissions)
+
+		remaining.forEach { expected ->
+			reporting.report(MissingRequiredActionPermissions, step) {
+				"${it} requires ${expected} permission for ${step.uses.action} to work."
+			}
+		}
+	}
+
+	private companion object {
+		val KnownActionPermissions: Map<String, Set<Scope>> = mapOf(
+				"actions/checkout" to setOf(Scope(Permission.CONTENTS, Access.READ)),
+		)
+
+		val MissingRequiredActionPermissions = Issue(
+				id = "MissingRequiredActionPermissions",
+				title = "Required permissions are not declared for action.",
+				description = """
+					Certain GitHub Actions require specific permissions to be specified for the GITHUB_TOKEN to work correctly.
+					
+					For example, the `actions/checkout` action requires `contents: read` permission to be able to
+					check out the repository.
+					
+					References:
+					
+				 	* [Documentation of `GITHUB_TOKEN` permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token)
+				 	* [List of Available permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token)
+				""".trimIndent(),
+				compliant = listOf(
+					Example(
+						explanation = "Required permissions are explicitly declared on the job.",
+						content = """
+							on: push
+							jobs:
+							  example:
+							    runs-on: ubuntu-latest
+							    permissions:
+							      contents: read
+							    steps:
+							      - uses: actions/checkout@v4
+						""".trimIndent(),
+					),
+				),
+				nonCompliant = listOf(
+					Example(
+						explanation = "Required permissions are not declared on the job or workflow.",
+						content = """
+							on: push
+							jobs:
+							  example:
+							    runs-on: ubuntu-latest
+							    permissions:
+							      pull-requests: write
+							    steps:
+							      - uses: actions/checkout@v4
+						""".trimIndent(),
+				),
```

Comment:
something is weird here, as if a closing `)` is missing

Reply:
Can you please check your Code Style? Does it say "Project"?
![image](https://github.com/TWiStErRob/net.twisterrob.ghlint/assets/2906988/ade1ba1b-cdb8-4fc4-8683-dfb9ff11d2b4)

Most of the code is "more" indented than normal.

Reply:
Note to self might need: `    <option name="PREFERRED_PROJECT_CODE_STYLE" value="Project" />`